code/game/machinery/autolathe/matterforge.dm

```diff
@@ -0,0 +1,594 @@
+		return 1
+
+	if(href_list["eject_material"] && (!current_design || paused || error))
+		var/material = href_list["eject_material"]
+		var/material/M = get_material_by_name(MATERIAL_COMPRESSED_MATTER)
+
+		if(!M.stack_type)
+			return
+
+		var/num = input("Enter sheets number to eject. 0-[stored_material[material]]","Eject",0) as num
+		if(!CanUseTopic(usr))
+			return
+
+		num = min(max(num,0), stored_material[material])
+		eject(num)
+		return 1
+
+
+	if(href_list["add_to_queue"])
+		var/recipe = href_list["add_to_queue"]
+		var/datum/design/picked_design
+
+		for(var/datum/design/f in design_list)
+			var/test = text2path((recipe))
+			if(f.id == test)
+				picked_design = f
+				break
+
+		if(picked_design)
+			var/amount = 1
+
+			if(href_list["several"])
+				amount = input("How many \"[picked_design.id]\" you want to print ?", "Print several") as null|num
+				if(!CanUseTopic(usr) || !(picked_design in design_list))
+					return
+
+			queue_design(picked_design, amount)
+
+		return 1
+
+	if(href_list["remove_from_queue"])
+		var/ind = text2num(href_list["remove_from_queue"])
+		if(ind >= 1 && ind <= queue.len)
+			queue.Cut(ind, ind + 1)
+		return 1
+
+	if(href_list["move_up_queue"])
+		var/ind = text2num(href_list["move_up_queue"])
+		if(ind >= 2 && ind <= queue.len)
+			queue.Swap(ind, ind - 1)
+		return 1
+
+	if(href_list["move_down_queue"])
+		var/ind = text2num(href_list["move_down_queue"])
+		if(ind >= 1 && ind <= queue.len-1)
+			queue.Swap(ind, ind + 1)
+		return 1
+
+
+	if(href_list["abort_print"])
+		abort()
+		return 1
+
+	if(href_list["pause"])
+		paused = !paused
+		return 1
+
+	if(href_list["unfold"])
+		if(unfolded == href_list["unfold"])
+			unfolded = null
+		else
+			unfolded = href_list["unfold"]
+		return 1
+
+/obj/machinery/matter_nanoforge/attackby(obj/item/I, mob/user)
+	GET_COMPONENT_FROM(comp, /datum/component/inspiration, I)
+	if(comp && comp.get_power() > 0)
+		user.drop_item(I)
+		I.forceMove(src)
+		if(power_source)
+			power_source.forceMove(loc)
+		power_source = I
+		return
+
+	if(power_source)
+		matter_assoc_list()	
+		eat(user, I)
+		return
+	else
+		to_chat(user, SPAN_WARNING("\The [src] does not have any artifact powering it."))
+
+/obj/machinery/matter_nanoforge/proc/eat(mob/living/user, obj/item/eating)
+	var/used_sheets
+	if(!eating && istype(user))
+		eating = user.get_active_hand()
+	if(!istype(eating))
+		return FALSE
+	if(stat)
+		return FALSE
+	if(stored_material[MATERIAL_COMPRESSED_MATTER] == storage_capacity)
+		return FALSE
+	if(!Adjacent(user) && !Adjacent(eating))
+		return FALSE
+	if(!length(eating.get_matter()))
+		to_chat(user, SPAN_WARNING("\The [eating] does not contain significant amounts of useful materials and cannot be accepted."))
+		return FALSE
+	var/total_used = 0     // Amount of material used.
+	var/mass_per_sheet = 0 // Amount of material constituting one sheet.
+	var/list/total_material_gained = list()
+	for(var/obj/O in eating.GetAllContents(includeSelf = TRUE))
```

Comment:
My worry here is that; how is the GetAllContents return structured?

If it returns the holding object first, and that is qdel'd, wouldn't that qdel the other items contained within it and reduce the matter return?